functions/signals/ks_combined.php

```diff
@@ -37,30 +37,94 @@
 	 */
 	public static function findState($tags, $next_speed, $next_speed_distant, $main_distance)
 	{
-		$state = "";
+		$state = ""; // start with unknown state
+		
+		// array with all states the signal can show
+		$possible_states = array();
+		
+		// check for coded tags
 		if(isset($tags["railway:signal:combined:states"]))
 		{
-			if ( $next_speed == 0 && strpos($tags["railway:signal:combined:states"], "hp0" )) // signal at end of route
+			// explode possible states
+			$possible_states_raw = explode(";", $tags["railway:signal:combined:states"]);
+			foreach($possible_states_raw as $curr_state)
 			{
-				$state = "hp0";
+				$namespace = explode(":", $curr_state);
+				// convert format "DE-ESO:state" to "state"
+				if($namespace[0] == "DE-ESO")
+				{
+					$possible_states[] = $namespace[1];
```

Comment:
Support people doing wrong tags like "DE-ESO:Hp0" CAN be done here, if wanted, by using strtolower